platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateSignatureCollector.java

```diff
@@ -0,0 +1,300 @@
+        Objects.requireNonNull(nodeId, "nodeId must not be null");
+        Objects.requireNonNull(signature, "signature must not be null");
+        final SignedState signedState = reservedSignedState.get();
+
+        if (signedState.addSignature(nodeId, signature)) {
+            // at this point the signed state is complete for the first time
+            signedStateMetrics.getStatesSignedPerSecondMetric().cycle();
+            signedStateMetrics
+                    .getAverageTimeToFullySignStateMetric()
+                    .update(Duration.between(signedState.getCreationTimestamp(), Instant.now())
+                            .toMillis());
+
+            return incompleteStates.remove(signedState.getRound());
+        }
+        return null;
+    }
+
+    /**
+     * Get the earliest round that is permitted to be stored in this data structure.
+     *
+     * @return the earliest round permitted to be stored
+     */
+    private long getEarliestPermittedRound() {
+        return lastStateRound - stateConfig.roundsToKeepForSigning() + 1;
+    }
+
+    /**
+     * Get rid of old states.
+     *
+     * @return a list of states that were purged
+     */
+    private @NonNull List<ReservedSignedState> purgeOldStates() {
+        final List<ReservedSignedState> purgedStates = new ArrayList<>();
+
+        // Any state older than this is unconditionally removed.
+        final long earliestPermittedRound = getEarliestPermittedRound();
+        for (final Iterator<ReservedSignedState> iterator =
+                        incompleteStates.values().iterator();
+                iterator.hasNext(); ) {
+            final ReservedSignedState reservedSignedState = iterator.next();
+            final SignedState signedState = reservedSignedState.get();
+            if (signedState.getRound() < earliestPermittedRound) {
+                signedStateMetrics.getTotalUnsignedStatesMetric().increment();
+                purgedStates.add(reservedSignedState);
+                iterator.remove();
+            }
+        }
+
+        signedStateMetrics.getUnsignedStatesMetric().update(incompleteStates.size());
+        return purgedStates;
+    }
+
+    /**
+     * Adjust the window where we are willing to save future signatures.
+     *
+     * @param currentRound the round of the most recently signed state
+     */
+    private void adjustSavedSignaturesWindow(final long currentRound) {
+        // Only save signatures for round N+1 and after.
+        // Any rounds behind this one will either have already had a SignedState
+        // added to this manager, or will never have a SignedState added to this manager.
+        if (savedSignatures.getFirstSequenceNumberInWindow() < currentRound + 1) {
+            savedSignatures.shiftWindow(currentRound + 1);
+        }
+    }
+
+    /**
+     * Clear the internal state of this collector.
+     *
+     * @param ignored ignored trigger object
+     */
+    public void clear(@NonNull final Object ignored) {
```

Comment:
There is no need to worry about answering, but I am curious about using this NonNull ignored parameter.

Reply:
This is really just a limitation of the wiring framework. I think that in the future we should be able to hide the ignored object and make it a detail hidden in the framework.